pkg/groupsync/manytomanysyncer.go

```diff
@@ -0,0 +1,181 @@
+		sourceGroupClient: sourceGroupClient,
+		targetGroupClient: targetGroupClient,
+		sourceGroupMapper: sourceGroupMapper,
+		targetGroupMapper: targetGroupMapper,
+		userMapper:        userMapper,
+	}
+}
+
+// SourceSystem returns the name of the source group system.
+func (f *ManyToManySyncer) SourceSystem() string {
+	return f.sourceSystem
+}
+
+// TargetSystem returns the name of the target group system.
+func (f *ManyToManySyncer) TargetSystem() string {
+	return f.targetSystem
+}
+
+// Sync syncs the source group with the given ID to the target group system.
+func (f *ManyToManySyncer) Sync(ctx context.Context, sourceGroupID string) error {
+	// get target group IDs for this source group ID
+	targetGroupIDs, err := f.sourceGroupMapper.MappedGroupIDs(ctx, sourceGroupID)
+	if err != nil {
+		return fmt.Errorf("error fetching target group IDs: %s, %w", sourceGroupID, err)
+	}
+
+	var merr error
+	for _, targetGroupID := range targetGroupIDs {
+		// get all source group IDs associated with the current target GroupID
+		sourceGroupIDs, err := f.targetGroupMapper.MappedGroupIDs(ctx, targetGroupID)
+		if err != nil {
+			merr = errors.Join(merr, fmt.Errorf("error getting associated source group ids: %w", err))
+			if len(sourceGroupIDs) == 0 {
+				// nothing left to do. move on to the next targetGroupID.
+				continue
+			}
+		}
+
+		// get the union of all users that are members of each source group
+		sourceUsers, err := f.sourceUsers(ctx, sourceGroupIDs)
+		if err != nil {
+			merr = errors.Join(merr, fmt.Errorf("error getting one or more source users: %w", err))
+			if len(sourceUsers) == 0 {
+				// nothing left to do. move on to the next targetGroupID.
+				continue
+			}
+		}
+
+		// map each source user to their corresponding target user
+		targetUsers, err := f.targetUsers(ctx, sourceUsers)
+		if err != nil {
+			merr = errors.Join(merr, fmt.Errorf("error getting one or more target users: %w", err))
+			if len(targetUsers) == 0 {
+				// nothing left to do. move on to the next targetGroupID.
+				continue
+			}
+		}
+
+		// map each targetUser to Member type
+		targetMembers := make([]Member, 0, len(targetUsers))
+		for _, user := range targetUsers {
+			targetMembers = append(targetMembers, &UserMember{Usr: user})
+		}
+
+		// targetMembers is now the canonical set of members for the target group ID.
+		// Set the target group's members to targetMembers.
+		if err := f.targetGroupClient.SetMembers(ctx, targetGroupID, targetMembers); err != nil {
+			merr = fmt.Errorf("error setting members to target group %s: %w", targetGroupID, err)
+		}
+	}
+
+	return merr
+}
+
+// SyncAll syncs all source groups that this GroupSyncer is aware of to the target system.
+func (f *ManyToManySyncer) SyncAll(ctx context.Context) error {
+	sourceGroupIDs, err := f.sourceGroupMapper.AllGroupIDs(ctx)
+	if err != nil {
+		return fmt.Errorf("error fetching source group IDs: %w", err)
+	}
+	var merr error
+	for _, sourceGroupID := range sourceGroupIDs {
```

Comment:
We might want to consider multi-threading such big batch sync

Reply:
Ok, I will make an issue for that and put in a TODO.

Reply:
https://github.com/abcxyz/team-link/issues/43